### PR TITLE
[refactor] Replace boost::function with std::function

### PR DESCRIPTION
**Pull request by yangzhg**
## Proposed changes

Replace boost::function with std::function

## Types of changes

What types of changes does your code introduce to Doris?
_Put an `x` in the boxes that apply_

- [ ] Bugfix (non-breaking change which fixes an issue)
- [ ] New feature (non-breaking change which adds functionality)
- [ ] Breaking change (fix or feature that would cause existing functionality to not work as expected)
- [ ] Documentation Update (if none of the other choices apply)
- [x] Code refactor (Modify the code structure, format the code, etc...)

## Checklist

_Put an `x` in the boxes that apply. You can also fill these out after creating the PR. If you're unsure about any of them, don't hesitate to ask. We're here to help! This is simply a reminder of what we are going to look for before merging your code._

- [ ] I have created an issue on (#5575) and described the bug/feature there in detail
- [ ] Compiling and unit tests pass locally with my changes
- [ ] I have added tests that prove my fix is effective or that my feature works
- [ ] If these changes need document changes, I have updated the document
- [ ] Any dependent changes have been merged

## Further comments

If this is a relatively large or complex change, kick off the discussion at dev@doris.apache.org by explaining why you chose the solution you did and what alternatives you considered, etc...

### CODE REVIEW

File: be/src/util/string_util.cpp [hunk 1]
```diff
@@ -18,14 +18,15 @@
...
+#include "util/hash_util.hpp"
 
 namespace doris {
 
 size_t hash_of_path(const std::string& identifier, const std::string& path) {
     size_t hash = std::hash<std::string>()(identifier);
     std::vector<std::string> path_parts = strings::Split(path, "/", strings::SkipWhitespace());
     for (auto& part : path_parts) {
-        boost::hash_combine<std::string>(hash, part);
+        HashUtil::hash_combine<std::string>(hash, part);
```
Comment (morningman) [inline be/src/util/string_util.cpp:29]: Is it exactly same as `boost::hash_combine`?

File: be/src/util/hash_util.hpp [hunk 2]
```diff
@@ -323,14 +337,18 @@ class HashUtil {
         murmur_hash3_x64_64(data, bytes, seed, &hash);
         return hash;
 #endif
-
+    }
+    template <class T>
```
Comment (morningman) [inline be/src/util/hash_util.hpp:341]: add comment to explain where does this code comes from

File: be/src/util/string_util.cpp (hunk 1 above)
Comment (yangzhg) [reply be/src/util/string_util.cpp:29]: it is the same with boost hash_combine

**Review by morningman — APPROVED** [review]
LGTM

**Pull request merged**